pkg/virt-handler/virt-chroot/virt-chroot.go

```diff
@@ -20,28 +20,65 @@
 package virt_chroot
 
 import (
+	"fmt"
+	"os"
 	"os/exec"
 	"strings"
 
+	ps "github.com/mitchellh/go-ps"
+
 	"kubevirt.io/kubevirt/pkg/safepath"
 	"kubevirt.io/kubevirt/pkg/unsafepath"
 )
 
 const (
-	binaryPath     = "/usr/bin/virt-chroot"
-	mountNamespace = "/proc/1/ns/mnt"
+	binaryPath            = "/usr/bin/virt-chroot"
+	MountNamespace        = "/proc/1/ns/mnt"
+	ContainerdProcessShim = "containerd-shim"
 )
 
+// recursively finding the ancestor process by matching the process nanme
+// returns an error if can not find
+func findParentProcessByName(ancestorProcess string) (ps.Process, error) {
+	pid := os.Getpid()
+	process, err := ps.FindProcess(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		if process.Executable() == ancestorProcess {
+			return process, nil
+		}
+		if process.PPid() == 0 {
+			break
+		}
+		process, err = ps.FindProcess(process.PPid())
+		if err != nil {
+			return nil, err
+		}
+	}
+	return nil, fmt.Errorf("failed to find the ancestor process: %s", ancestorProcess)
+}
+
+func GetHostNamespacePath() string {
+	process, err := findParentProcessByName(ContainerdProcessShim)
```

Comment:
This would mean that Kubevirt should work only with containerd, or am I missing anything?

Reply:
No, if you look at this function findParentProcessByName, if it finds the upstream parent has the 'containerd-shim' name, then the process is valid and pid will be used for the /proc/pid/ns/mnt, otherwise the /proc/1/ns/mnt will be used here as the original intent of the code. 

The parent processes in the case of kubevirt/k8s under container is like this 1) init
2) containerd-shim 3) containerd-shim 4) name of the vm being launched;
In this case, we will find the process id in #3, and the namespace under that
process of the mount is correct for the k8s and kubevirt file systems;

In the case of non-container, it will be like 1) init 2) containerd-shim 3) name of the vm launched.
when it hits #2, that is the kubevirt process, which is in the same namespace of the 'init' process
in mounting the filesystem. If it does not find the name of 'container-shim', then it will go to init,
which is the original code is doing.

Reply:
If the runtime runs in host or VM, but not inside another container, it should not be affected.
Here is the "ps aux | grep container", the 'kube' container has the k3s/kubevirt running:

07ab2c40-408a-4bc2-b9f2-8ca94235074f:/# ps aux | grep container
  842 root      0:25 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 5fd3bbcc47a275e789711dd916bbcb848ca552e92c89a969f67d50d73dec117c -address /run/k3s/containerd/containerd.sock
 1319 root      0:30 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id b19f4ac3bc599092503ce9984abc6bc754b6ad4d4bc8dcfbb90c9a90b973951f -address /run/k3s/containerd/containerd.sock
 1416 root      0:25 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id efe5ed7a1d7598ef31282ea04318c4751c14e480b05a9e720ea3874f1b448c5c -address /run/k3s/containerd/containerd.sock
 1426 root      0:28 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 521e6ca756cde4bcfc32bbc246ea3b615ba9c4ff4078c81558908bda3e38540c -address /run/k3s/containerd/containerd.sock
 1642 root      1:44 /usr/bin/containerd
 1663 root      0:02 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id debug -address /run/containerd/containerd.sock
 1704 root      0:02 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id edgeview -address /run/containerd/containerd.sock
 1748 root      0:02 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id guacd -address /run/containerd/containerd.sock
 1788 root      0:02 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id kube -address /run/containerd/containerd.sock  <=== this 'kube' container, runs the k3s/kubervernetes/kubevirt ====
 1848 root      0:02 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id newlogd -address /run/containerd/containerd.sock
 1889 root      1:28 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id pillar -address /run/containerd/containerd.sock
 1991 root      0:02 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id vtpm -address /run/containerd/containerd.sock
 2421 root      0:02 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id watchdog -address /run/containerd/containerd.sock
 2785 root      0:02 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id wlan -address /run/containerd/containerd.sock
 3100 root      0:02 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id wwan -address /run/containerd/containerd.sock
 3384 root      0:02 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id xen-tools -address /run/containerd/containerd.sock
 4009 root      0:28 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 86968ae6e431a34958a1fcbe5d04210976b45042b5c4c9d7ab1f1e471db9146b -address /run/k3s/containerd/containerd.sock
 4217 root      0:27 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 349367896334867ea142b8949ff5bc2c24a3ab44ab3e17a8919b639b223d84fb -address /run/k3s/containerd/containerd.sock
 4329 root      0:26 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id e1eb0c6ea41fa93044ffdff27a6386936d30494b416ff110511708fdfd993bba -address /run/k3s/containerd/containerd.sock
 4636 root      0:28 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 2d60caa7c7679741701dd4a50712d5197ddfedcc3e0b42a89bfcd477f25fb0d7 -address /run/k3s/containerd/containerd.sock
 4639 root      0:27 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 23e6ff05a4ecc7a9859d845c51bb8a91294adc161a2550c76897cf5593406efe -address /run/k3s/containerd/containerd.sock
 4678 root      0:29 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 24f19ab66f46c958a7674810de37f88a694b078575b634cb044587974c4d330e -address /run/k3s/containerd/containerd.sock
10580 root      0:00 grep container
11757 root     18:02 containerd -c /var/lib/rancher/k3s/agent/etc/containerd/config.toml -a /run/k3s/containerd/containerd.sock --state /run/k3s/containerd --root /var/lib/rancher/k3s/agent/containerd
23149 root      0:27 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 439b20ec55bc32ac60cd4a2bb8e52a253f992fcc3ba498bfaa9cca340b6ff2dc -address /run/k3s/containerd/containerd.sock
24490 root      0:45 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 5ad7d683a7588709e5d8341b120ebd35ad3f9937e662732e1109ded26ebacb63 -address /run/k3s/containerd/containerd.sock
25086 107       0:00 /usr/bin/virt-launcher-monitor --qemu-timeout 270s --name local-pvc-vm --uid 972ca801-ac21-4432-83c6-8a836bfd515f --namespace eve-kube-app --kubevirt-share-dir /var/run/kubevirt --ephemeral-disk-dir /var/run/kubevirt-ephemeral-disks --container-disk-dir /var/run/kubevirt/container-disks --grace-period-seconds 45 --hook-sidecars 0 --ovmf-path /usr/share/OVMF --run-as-nonroot
25281 107       0:42 /usr/bin/virt-launcher --qemu-timeout 270s --name local-pvc-vm --uid 972ca801-ac21-4432-83c6-8a836bfd515f --namespace eve-kube-app --kubevirt-share-dir /var/run/kubevirt --ephemeral-disk-dir /var/run/kubevirt-ephemeral-disks --container-disk-dir /var/run/kubevirt/container-disks --grace-period-seconds 45 --hook-sidecars 0 --ovmf-path /usr/share/OVMF --run-as-nonroot
26090 107       0:04 /usr/bin/container-disk --copy-path /var/run/kubevirt-ephemeral-disks/container-disk-data/972ca801-ac21-4432-83c6-8a836bfd515f/disk_0
26644 root      0:27 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 9568a5eb45f8523d315485ad5ee5fb219634cb29fddce39f1bcc0388e8d46d42 -address /run/k3s/containerd/containerd.sock
26717 root      0:45 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 9149d081fc2a8ecd917c6811bb05e0b6ec37e9c9730c6c82d5e8e1a751a16e66 -address /run/k3s/containerd/containerd.sock
28216 root      0:27 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id ac2321a34f7339641ffee3dc911e352b01cd0433f01416256c4400785406fd7e -address /run/k3s/containerd/containerd.sock
29245 root      0:26 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id d17b6a17995846e14050ec1ef348001108db29344a2d97731c73cddb60296ca0 -address /run/k3s/containerd/containerd.sock
29757 root      1:41 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id 6c9b6d54930d02ad666fce0f0d22bd6c4c0339627d354d229338b13140bb908c -address /run/k3s/containerd/containerd.sock
29876 root      0:26 /var/lib/rancher/k3s/data/c26e7571d760c5f199d18efd197114f1ca4ab1e6ffe494f96feb65c87fcb8cf0/bin/containerd-shim-runc-v2 -namespace k8s.io -id f30e04f7cf80c2bf1037a7f46e4fe30f472ca49a154177030a94f0ec0a4a2e87 -address /run/k3s/containerd/containerd.sock
07ab2c40-408a-4bc2-b9f2-8ca94235074f:/#

Reply:
Do I understand correctly that this is the output of `ps aux` inside the container that acts as a Kubernetes node?
Can you also share the output of `virt-handler`?

Reply:
@xpivarc this is correct, above output is run inside the 'container' which is the kubernetes node (runs k3s server).

If I run the logs (with the setcap, otherwise the init container keeps crashing):
[virt-handler.log](https://github.com/kubevirt/kubevirt/files/12011367/virt-handler.log)

Reply:
@xpivarc @vasiliy-ul @alicefr lets come to the other part of the PR :-)

We are not the only one uses this approach, please see this from longhorn, they are implementing the same way:
[longhorn-code](https://github.com/longhorn/go-spdk-helper/blob/main/pkg/util/process.go#L61).

we also independently committed the same solution to l[onghorn-PR-#63](https://github.com/longhorn/go-iscsi-helper/pull/63), to make longhorn working under the container based kubernetes node.

Reply:
@naiming-zededa, searching for `containerd-shim` process does not seem to be a generic solution (as @xpivarc already mentioned). It will work only if you use `containerd`. In case you have the same containerized setup but with another container runtime (e.g. `cri-o`, `docker`, or smth else) this, I guess, will not work because the process you need to search for will have different name.

Ideally, it would be best if we could figure out an approach that works with **any** container runtime. Though it is still under question if that is achievable. Let's maybe try to figure out a more generic solution. Could you share the **full** outputs of `ps aux`: 1) from the host; 2) from the node container; 3) from virt-handler pod? I am not saying that we will find smth there right away, but let's have a look.

Reply:
@vasiliy-ul here is the output of those three places:
[kubevirt-process.txt](https://github.com/kubevirt/kubevirt/files/12052999/kubevirt-process.txt)

Reply:
pstree output:
[kubevirt-pstree.txt](https://github.com/kubevirt/kubevirt/files/12053022/kubevirt-pstree.txt)

Reply:
Here is a output of using 'virt-chroot' for different namespace mount for various chain of processes,
1) init -> 2) 'kube' container -> 3) k3s launched kubevirt -> 4) virt-handler
we can see that only the step 3 using the 'virt-chroot' generates the correct /var/lib directory in this case.
a file is attached for the output:
[kubevirt-chroot.txt](https://github.com/kubevirt/kubevirt/files/12053561/kubevirt-chroot.txt)

If we don't want to hardcode the container names, one way also works is that, instead of
looping in for {} to find the container name, we return immediately for the parent process of the caller.
I think it will work for non-container/VM based case too

Reply:
Right, seems that the parent process should also have the correct mount namespace. I checked on a VM cluster, and it refers to the same as init:

```console
node21:~ # pgrep virt-handler
25527
node21:~ # grep PPid /proc/25527/status 
PPid:	25289
node21:~ # ls -la /proc/25289/ns/mnt
lrwxrwxrwx 1 root root 0 Jul 17 06:15 /proc/25289/ns/mnt -> mnt:[4026531840]
node21:~ # ls -la /proc/1/ns/mnt
lrwxrwxrwx 1 root root 0 Jul 17 06:15 /proc/1/ns/mnt -> mnt:[4026531840]
```

Alternatively, I guess that **any** process running in the node container (or on the host system with the regular non-containerized KubeVirt setup) will have the needed mount namespace. I was thinking about `kubelet`:

```console
node21:~ # ls -la /proc/$(pgrep kubelet)/ns/mnt
lrwxrwxrwx 1 root root 0 Jul 17 06:15 /proc/1659/ns/mnt -> mnt:[4026531840]
```

But using the parent process sounds like easier, and I currently do not see potential problems with that.

Reply:
@vasiliy-ul ok, I'll update this PR with this using the parent process.